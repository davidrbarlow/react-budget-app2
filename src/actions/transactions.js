import axios from 'axios';
import { removeSelectedTransactionIds } from './selectedRows';


export const addTransaction = (transaction) => ({
  type: 'ADD_TRANSACTION',
  transaction
});

export const startAddTransaction = (transactionData = {}) => {

  return (dispatch, getState) => {
    const token = getState().auth.authToken;
    const {
        description = '', 
        amount = 0, 
        postedAt = 0 ,
        cycle = ''
    } = transactionData;
  
    const transaction = {description, amount, postedAt, cycle};

    //add transaction api
    console.log('start add trans ', `${process.env.REACT_APP_API_URL}/transaction/transaction`);
    return axios.post(`${process.env.REACT_APP_API_URL}/transaction/transaction`,
      transaction,
      {
        headers : { 'x-auth':token}
      }
    ).then((res)=>{
    }).catch((e)=>{
      console.log(e);
    });
  };
};

export const editTransaction = (id, updates) =>({
  type: 'EDIT_TRANSACTION',
  id,
  updates
});

//startEditTransaction

export const startEditTransaction = (id, updates) => {
  return (dispatch,getState) => {
   const token = getState().auth.authToken;
   console.log(token);
  return  axios.patch(`${process.env.REACT_APP_API_URL}/transaction/edit/${id}`, 
  updates,
  {
    headers: {
      'x-auth': token
    }
  })
  .then(()=>{   
    dispatch(editTransaction(id, updates));
    });
  };
}

export const removeTransaction = (id) => ({
  type: 'REMOVE_TRANSACTION',
  id
});

export const startRemoveTransaction = (id) => {
  console.log('ID', id);
  return (dispatch,getState) => {
    const token = getState().auth.authToken;
    return  axios.delete(`${process.env.REACT_APP_API_URL}/transaction/remove/${id}`,
    {
      headers: {
        'x-auth': token
      }
    })
    .then(()=>{   
      dispatch(removeTransaction(id));
      });
   };
}

export const removeTransactions = (ids) => ({
  type: 'REMOVE_TRANSACTIONS',
  ids
});

export const startRemoveTransactions = (ids) => {
 // const idJSON = JSON.stringify({ids});
  return (dispatch,getState) => {
    const token = getState().auth.authToken;
    return  axios.patch(`${process.env.REACT_APP_API_URL}/transaction/removeSelected`,
    {
      ids
    },
    {
      headers: {'x-auth': token}
    })
    .then(()=>{   
      dispatch(removeTransactions(ids));
      dispatch(removeSelectedTransactionIds(ids));
    });
   };
}



export const setTransactions = (transactions) => ({
  type:'SET_TRANSACTIONS',
  transactions
});

export const startSetTransactions = () => {
  
  return (dispatch, getState)=>{
    const token = getState().auth.authToken;
    return axios.get(`${process.env.REACT_APP_API_URL}/transaction`,
    {
      headers:{
        'x-auth':token
      }
    }
    ).then((res)=>{
    dispatch(setTransactions(res.data.transactions));
  })
  }

};




export const csvUpload = (file, token) => {
    return axios.post(`${process.env.REACT_APP_API_URL}/transaction/upload`,
      file,
      { headers: {
        'Content-Type': 'multipart/form-data',
        'x-auth': token
        }
      },
      {
        onUploadProgress: ProgressEvent => {
          this.setState({
            loaded: (ProgressEvent.loaded / ProgressEvent.total*100),
          })
        },
      })
      .then((res)=>{
        if(res.data.status === 400){
          throw(res);
        }else{
          return(res);
        }  
      })
      // .catch((e)=>{
      //   console.log('axios e', e);
      //   return Promise.reject(e);
      // })
};

