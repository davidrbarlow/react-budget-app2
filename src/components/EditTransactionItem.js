import React from 'react';
import {connect} from 'react-redux';
import TransactionItemForm from './TransactionItemForm';

const EditTransactionItem = (props) => {

  return (
    <div>
      <TransactionItemForms
        transaction = {props.transaction}
        onSubmit={(transaction)=>{
        props.dispatch(startEditTransaction(props.transaction.id, transaction));
        console.log(props.match.params.id);
      }}
      />


    </div>
  )
}

const mapStateToProps = (state, props) => {
  return {
      expense: state.transctions.find((transaction)=>transaction.id === props.match.params.id)
  };
};
export default connect(mapStateToProps)(EditExpensePage);