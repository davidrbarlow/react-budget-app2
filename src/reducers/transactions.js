const transactionsReducerDefaultState = [];

export default (state = transactionsReducerDefaultState, action)=>{
  switch (action.type){
    case 'ADD_TRANSACTION':
      return [
          ...state,
      action.transaction];
    case 'REMOVE_TRANSACTION':
      return state.filter((transaction)=>{return action.id !== transaction._id });
    case 'REMOVE_TRANSACTIONS':
      return state.filter((transaction)=>{return !action.ids.includes(transaction._id)});
    case 'EDIT_TRANSACTION':
      return state.map((transaction)=>{
        if (transaction.id === action.id){
          return {
            ...transaction,
            ...action.updates
          };
        } else {
          return transaction;
        };
      });
    case 'SET_TRANSACTIONS':
        return action.transactions;
    default:
        return state;
  }


}