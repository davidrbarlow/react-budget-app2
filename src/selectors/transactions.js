import moment from 'moment';
//Get visible expenses

let adjustedTransactions = [];

export default (transactions, filters) => {
  
    setCycleBalance(transactions);
    return filterTransactions(transactions, filters);
};

const setCycleBalance = (transactions) => {
    transactions.forEach((transaction)=>{
        if (transaction.setAsBalance) {
            transaction.cycle = 'Balance';  
        }
        adjustedTransactions.push(transaction);
      
    });
};
//setCycleBalance();

const filterTransactions =  (transactions, {text, sortBy, startDate, endDate}) => {
    return transactions.filter((transaction)=>{
        const postedAtMoment = moment(transaction.postedAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(postedAtMoment, 'day'): true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(postedAtMoment, 'day'): true;
        const textMatch = transaction.description.toLowerCase().includes(text.toLowerCase());
        const projectionNoMatch = transaction.accountType !== 'Projection';
       // console.log('transaction from selectors', transaction);
        return startDateMatch && endDateMatch && textMatch && projectionNoMatch;

    }).sort((a,b)=>{
        if (sortBy === 'date'){
            return a.postedAt<b.postedAt ? 1 : -1;
        }
        if (sortBy === 'amount') {
          //  return a.amount < b.amount ? 1 : -1
           return  b.amount - a.amount; 
        }
        else{
          return a.postedAt<b.postedAt ? 1 : -1;
        }
    });
};