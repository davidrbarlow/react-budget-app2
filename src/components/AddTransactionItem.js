import React from 'react';
import {connect} from 'react-redux';
import { Route , withRouter} from 'react-router-dom';
import TransactionItemForm from './TransactionItemForm';
import {startAddTransaction}  from '../actions/transactions';
import {toggleAddTransaction}  from '../actions/pageEdits';

export class AddTransactionItem extends React.Component {
 
  constructor(props) {
    super(props);
  
  }

    onSubmit = async (transaction) => {
      console.log('addtransitem', this.props);
        await this.props.startAddTransaction(transaction);
        // window.location.reload();
        await this.props.handleState();
        console.log('props in on submit ',this.props);
        this.props.toggleAddTransaction();
    };

    render() {
        return(
        <div>
            <TransactionItemForm 
            onSubmit={this.onSubmit}
            />
        </div>
    );
    }

}


const mapDispatchToProps = (dispatch) => ({
  startAddTransaction: (transaction) => dispatch(startAddTransaction(transaction)),
  toggleAddTransaction: () => dispatch(toggleAddTransaction())
})

export default connect(undefined, mapDispatchToProps)(withRouter(AddTransactionItem));