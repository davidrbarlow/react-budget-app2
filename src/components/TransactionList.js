import React from 'react';
import {connect} from 'react-redux';
import TransactionListItem from './TransactionListItem';
import selectTransactions from '../selectors/transactions';
import selectProjections from '../selectors/projections';
import AddTransactionItem  from './AddTransactionItem';
import { startSetTransactions } from '../actions/transactions';




export class TransactionList extends React.Component {
//export const TransactionList = (props) => (
    constructor(props) {
        super(props);

        this.handleState = this.handleState.bind(this);
      
      }

 handleState = async () => {
     await this.props.startSetTransactions();
     await this.setState(this.state);
    };
    
render(){

    const heading = (this.props.pageEdits.activePage === 'dashboard') ? 'Transaction' : 'Projection';
    
    let list;
    
    if (this.props.transactions.length === 0) {
        list = <div>
                    <span>No Transactions</span>
                </div>
    } 
     else if (this.props.pageEdits.activePage === 'dashboard'){
         list = this.props.transactions.map((transaction) =>  
        <TransactionListItem key={transaction._id} {...transaction} handleState={this.handleState}/> 
         )
        //  list = this.props.transactions.map((transaction) =>  
        // <TransactionListItem key={transaction._id} {...transaction} handleState={this.handleState}/> 
     }
     else if (this.props.pageEdits.activePage === 'projection'){
        list = this.props.projections.map((projection) =>  
       <TransactionListItem key={projection._id} {...projection} handleState={this.handleState}/> 
        )
     }

    return (
    <div>
    <div className="grid-container" >
        
        <span>
            <div className="transaction-header">{heading}</div>
        </span>
        <div>
            {this.props.pageEdits.addTransaction && <AddTransactionItem handleState={this.handleState}/>}    
        </div>
         <div>
         {list}
         </div>
            
    </div>
    </div>
);
//handleState={ this.handleState }
}
 }


 const mapDispatchToProps = (dispatch) => ({
    startSetTransactions : (token) => dispatch(startSetTransactions(token)),
  //  toggleAddTransaction : () => dispatch(toggleAddTransaction())
    })
 
const mapStateToProps = (state) => {
  console.log('mapStateToProps state ', state);
    return {
      transactions: selectTransactions(state.transactions, state.filters),
      projections: selectProjections(state.transactions, state.filters),
      pageEdits: state.pageEdits,
    };
}
 
 
export default connect(mapStateToProps,mapDispatchToProps)(TransactionList);

// {
//     this.props.transactions.length === 0 ? (
//          <div>
//              <span>No Transactions</span>
//          </div>
//      ):(
//          this.props.transactions.map((transaction) =>  
//          <TransactionListItem key={transaction._id} {...transaction} handleState={this.handleState}/>
//          )
//      )
     
//  }