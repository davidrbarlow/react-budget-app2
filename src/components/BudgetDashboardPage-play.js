import React from 'react';
import {connect} from 'react-redux';
import NavIcons from './NavIcons';
import Header from './Header';
import TransactionList from './TransactionList';
import TransactionListFilters from './TransactionListFilters';
import TransactionListControls from './TransactionListControls';

import { logout } from '../actions/auth';


class BudgetDashboardPage extends React.Component {

    constructor(props) {
        super(props);
        this.onUnload = this.onUnload.bind(this); // if you need to bind callback to this
    }

    componentWillUnmount() {
        // Make sure to remove the DOM listener when the component is unmounted.
        this.nv.removeEventListener("nv-enter", this.handleNvEnter);
      }

      componentDidMount() {
        // When the component is mounted, add your DOM listener to the "nv" elem.
        // (The "nv" elem is assigned in the render function.)
        this.nv.addEventListener("nv-enter", this.handleNvEnter);
      }

      handleNvEnter = (event) => {
        console.log("Nv Enter:", event);
      }


    onUnload = (event) => {
        console.log("hellooww");
        event.returnValue = "Hellooww";
        this.props.logout(this.props.token);
    }

    render() {
        return (
            <div className="layout">
                <div className="layout--top-left-corner"></div>
                <div className="layout--header">
                    <Header />
                </div>
                <div className="layout--nav-icon">
                    <NavIcons />
                </div>
                <div className="layout--filters">
                    <TransactionListFilters />
                </div>
            
                <div className="layout--controls">
                    <TransactionListControls />
                </div>
                <div className="layout--transactions">
                    <TransactionList />
                </div>       
            </div>
        );
    }
}

const mapDispatchToProps = (dispatch) => {
    return{
        logout: (token) => dispatch(logout(token))
    };
} ;

const mapStateToProps = (state) => {
    return {
        authToken: state.auth.authToken,
    }
}


export default connect(mapStateToProps, mapDispatchToProps)(BudgetDashboardPage);