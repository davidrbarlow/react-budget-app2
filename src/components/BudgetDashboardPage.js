import React from 'react';
import NavIcons from './NavIcons';
import Header from './Header';
import TransactionList from './TransactionList';
import TransactionListFilters from './TransactionListFilters';
import TransactionListControls from './TransactionListControls';


const BudgetDashboardPage = () => (
    <div className="layout">
        <div className="layout--top-left-corner"></div>
        <div className="layout--header">
            <Header />
        </div>
        <div className="layout--nav-icon">
            <NavIcons />
        </div>
        <div className="layout--filters">
            <TransactionListFilters />
        </div>
       
        <div className="layout--controls">
             <TransactionListControls />
        </div>
        <div className="layout--transactions">
            <TransactionList />
        </div>
        
        
        
    </div>
);

export default BudgetDashboardPage;