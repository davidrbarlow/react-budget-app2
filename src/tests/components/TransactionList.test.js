import React from 'react';
import { shallow } from 'enzyme';
import { TransactionList } from '../../components/TransactionList';
import transactions from '../fixtures/transactions';
//import { wrap } from 'module';
import toJSON from 'enzyme-to-json';


const pageEdits = {
  addTransaction : false,
};

test('should render TransactionList with Transcations ',() => {
  const wrapper = shallow(<TransactionList transactions={transactions} pageEdits={pageEdits} />);
  expect(toJSON(wrapper)).toMatchSnapshot();
});

test('should display add transaction row', () => {
  const pageEdits = {
    addTransaction : true,
  };

  const wrapper = shallow(<TransactionList transactions={transactions} pageEdits={pageEdits} />);
  expect(toJSON(wrapper)).toMatchSnapshot();
 // wrapper.find('span').at(0).to.equal('No Transactions');
  
});

test('should display no transactions if no transactions', () => {
  const transcations = [];
  const wrapper = shallow(<TransactionList transactions={transcations} pageEdits={pageEdits} />);
  expect(toJSON(wrapper)).toMatchSnapshot();
 // wrapper.find('span').at(0).to.equal('No Transactions');
  
});



