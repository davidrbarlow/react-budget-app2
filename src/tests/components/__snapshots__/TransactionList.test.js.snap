// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display add transaction row 1`] = `
<div>
  <div
    className="grid-container"
  >
    <span>
      <div
        className="transaction-header"
      >
        Transactions
      </div>
    </span>
    <div>
      <Connect(AddTransactionItem)
        handleState={[Function]}
      />
    </div>
    <Connect(TransactionListItem)
      _id="5c705b052c747592ec78279d"
      accountType="Bank"
      amount={4000}
      balance={8111.86}
      description="Online Transfer from CHK ...3025 transaction#: 7899743059"
      handleState={[Function]}
      key="5c705b052c747592ec78279d"
      postedAt={0}
    />
    <Connect(TransactionListItem)
      _id="5c705b052c747592ec78279a"
      accountType="Bank"
      amount={-1096.34}
      balance={3017.1}
      description="USBANK LOAN      PAYMENT                    PPD ID:  580580574"
      handleState={[Function]}
      key="5c705b052c747592ec78279a"
      postedAt={-345600000}
    />
    <Connect(TransactionListItem)
      _id="5c705b052c747592ec78279c"
      accountType="Bank"
      amount={-140}
      balance={7971.86}
      description="Online Payment 7899735230 To AT&T MOBILITY 02/01"
      handleState={[Function]}
      key="5c705b052c747592ec78279c"
      postedAt={345600000}
    />
  </div>
</div>
`;

exports[`should display no transactions if no transactions 1`] = `
<div>
  <div
    className="grid-container"
  >
    <span>
      <div
        className="transaction-header"
      >
        Transactions
      </div>
    </span>
    <div />
    <div>
      <span>
        No Transactions
      </span>
    </div>
  </div>
</div>
`;

exports[`should render TransactionList with Transcations  1`] = `
<div>
  <div
    className="grid-container"
  >
    <span>
      <div
        className="transaction-header"
      >
        Transactions
      </div>
    </span>
    <div />
    <Connect(TransactionListItem)
      _id="5c705b052c747592ec78279d"
      accountType="Bank"
      amount={4000}
      balance={8111.86}
      description="Online Transfer from CHK ...3025 transaction#: 7899743059"
      handleState={[Function]}
      key="5c705b052c747592ec78279d"
      postedAt={0}
    />
    <Connect(TransactionListItem)
      _id="5c705b052c747592ec78279a"
      accountType="Bank"
      amount={-1096.34}
      balance={3017.1}
      description="USBANK LOAN      PAYMENT                    PPD ID:  580580574"
      handleState={[Function]}
      key="5c705b052c747592ec78279a"
      postedAt={-345600000}
    />
    <Connect(TransactionListItem)
      _id="5c705b052c747592ec78279c"
      accountType="Bank"
      amount={-140}
      balance={7971.86}
      description="Online Payment 7899735230 To AT&T MOBILITY 02/01"
      handleState={[Function]}
      key="5c705b052c747592ec78279c"
      postedAt={345600000}
    />
  </div>
</div>
`;
